### PR TITLE
【Inplace api】Add copy for inplace

### PR DESCRIPTION
**Pull request by GGBond8488**
### PR types
Others

### PR changes
Others 

### Description

Pcard-72375
add Input copy for inpalce api backward calculate

#### Normal api change

Before:
 ```
paddle::Tensor pow_ad_func(const paddle::Tensor& x, paddle::experimental::Scalar y) {
...
  // Get Input AutoGradMeta
  egr::AutogradMeta* x_autograd_meta = egr::EagerUtils::nullable_autograd_meta(x);
...
  // Forward API Call
  auto api_result = paddle::experimental::pow(x, y);
...
...
  // Get Outputs
  auto& out = api_result;

  // Get Output AutoGradMeta
  egr::AutogradMeta* out_autograd_meta = egr::EagerUtils::autograd_meta(&out);
  bool trace_backward = egr::Controller::Instance().HasGrad();
  bool require_any_grad = egr::EagerUtils::ComputeRequireGrad(trace_backward,x_autograd_meta);

  // Check Inplace if needed

  // Node Creation
  if(require_any_grad) {
...
    // Node Construction
    auto grad_node = std::shared_ptr<PowGradNode>(new PowGradNode(1, 1));
    // Set for forward trace
  if (FLAGS_check_nan_inf) {
    grad_node->SetForwardTrace(egr::Controller::Instance().GetPythonStack());
  }
    // SetAttributes if needed
    grad_node->SetAttributefactor(factor);
    // Set TensorWrappers for Forward Inputs if needed
    grad_node->SetTensorWrapperx(x);
    // SetGradOutMeta & SetEdges
    grad_node->SetGradOutMeta(x, 0);
    // SetOutRank & SetHistory & SetGradInMeta
    if (out_autograd_meta) {
      egr::EagerUtils::SetOutRankWithSlot(out_autograd_meta, 0);
    }
    if (out_autograd_meta) {
      egr::EagerUtils::SetHistory(out_autograd_meta, grad_node);
    }
    grad_node->SetGradInMeta(out, 0);
    // Set TensorWrappers for Forward Outputs if needed

  }
```

After this pr merge:
```
paddle::Tensor pow_ad_func(const paddle::Tensor& x, paddle::experimental::Scalar y) {
...
  // Get Input AutoGradMeta
  egr::AutogradMeta* x_autograd_meta = egr::EagerUtils::nullable_autograd_meta(x);

  bool trace_backward = egr::Controller::Instance().HasGrad();
  bool require_any_grad = egr::EagerUtils::ComputeRequireGrad(trace_backward,x_autograd_meta);

  // Node Declaration
  std::shared_ptr<PowGradNode> grad_node;

  // Set grad_node before API Call
  if(require_any_grad) {
    paddle::platform::RecordEvent node_creation_record_event("pow node_creation", paddle::platform::TracerEventType::OperatorInner, 1);
    // Node Construction
    grad_node = std::shared_ptr<PowGradNode>(new PowGradNode(1, 1));
    // Set for forward trace
  if (FLAGS_check_nan_inf) {
    grad_node->SetForwardTrace(egr::Controller::Instance().GetPythonStack());
  }
    // SetAttributes if needed
    grad_node->SetAttributey(y);
    // Set TensorWrappers for Forward Inputs if needed
    grad_node->SetTensorWrapperx(x);
  }

 // Forward API Call
  auto api_result = paddle::experimental::pow(x, y);
...
  // Get Outputs
  auto& out = api_result;

  // Get Output AutoGradMeta
  egr::AutogradMeta* out_autograd_meta = egr::EagerUtils::autograd_meta(&out);
...
  // Set grad_node after API call
  if(require_any_grad) {
    egr::EagerUtils::PassStopGradient(false,out_autograd_meta);
    // SetGradOutMeta & SetEdges
    grad_node->SetGradOutMeta(x, 0);
    // SetOutRank & SetHistory & SetGradInMeta
    if (out_autograd_meta) {
      egr::EagerUtils::SetOutRankWithSlot(out_autograd_meta, 0);
    }
    if (out_autograd_meta) {
      egr::EagerUtils::SetHistory(out_autograd_meta, grad_node);
    }
    grad_node->SetGradInMeta(out, 0);
    // Set TensorWrappers for Forward Outputs if needed
  }

  VLOG(4) << "Finish AD API: pow";`
```
#### Inplace api change
will add Input copy in inplace pow as follow

```
paddle::Tensor pow__ad_func(const paddle::Tensor& x, paddle::experimental::Scalar y) {
...
  // Set grad_node before API Call
  if(require_any_grad) {
    paddle::platform::RecordEvent node_creation_record_event("pow node_creation", paddle::platform::TracerEventType::OperatorInner, 1);
    // Node Construction
    grad_node = std::shared_ptr<PowGradNode>(new PowGradNode(1, 1));
    // Set for forward trace
  if (FLAGS_check_nan_inf) {
    grad_node->SetForwardTrace(egr::Controller::Instance().GetPythonStack());
  }
    // SetAttributes if needed
    grad_node->SetAttributey(y);
    // Set TensorWrappers for Forward Inputs if needed
    auto x_clone = paddle::experimental::assign(x);
    grad_node->SetTensorWrapperx(x_clone);
  }

 // Forward API Call
  auto api_result = paddle::experimental::pow(x, y);
  // Check NaN and Inf if needed

  // Get Outputs
  auto& out = api_result;
...
```

### CODE REVIEW

File: paddle/fluid/eager/auto_code_generator/generator/eager_gen.py [hunk 1]
```diff
@@ -226,6 +226,16 @@ class {} : public egr::GradNodeBase {{
   VLOG(5) << \"Running C++ API: \" << \"{}\";
  // Before log info
 {}
+
+  bool trace_backward = egr::Controller::Instance().HasGrad();
+  bool require_any_grad = egr::EagerUtils::ComputeRequireGrad({});
+
+ // Node Declaration
```
Comment (chenwhql) [inline paddle/fluid/eager/auto_code_generator/generator/eager_gen.py]: 生成代码的缩进也需要对齐

File: paddle/fluid/eager/auto_code_generator/generator/eager_gen.py [hunk 2]
```diff
@@ -226,6 +226,16 @@ class {} : public egr::GradNodeBase {{
...
+  std::shared_ptr<{}> grad_node;
+
+  //Set grad_node before API Call
```
Comment (chenwhql) [inline paddle/fluid/eager/auto_code_generator/generator/eager_gen.py]: 注释后需要一个空格

File: paddle/fluid/eager/auto_code_generator/generator/eager_gen.py [hunk 3]
```diff
@@ -234,12 +244,9 @@ class {} : public egr::GradNodeBase {{
 {}
   // Get Output AutoGradMeta
 {}
-  bool trace_backward = egr::Controller::Instance().HasGrad();
-  bool require_any_grad = egr::EagerUtils::ComputeRequireGrad({});
-
   // Check Inplace if needed
 {}{}
-  // Node Creation
+  //Set grad_node after API call
```
Comment (chenwhql) [inline paddle/fluid/eager/auto_code_generator/generator/eager_gen.py]: 同上

File: test/legacy_test/test_pow.py [hunk 4]
```diff
@@ -213,5 +214,40 @@ def test_errors(self):
...
+        return paddle.pow(var, self.y)
+
+    def test_type_error(self):
+        var = paddle.to_tensor(self.input_var_numpy, dtype=self.dtype)
+        with self.assertRaisesRegex(
+            TypeError,
+            'y must be scalar or tensor type, but received: %s ' % (type([2])),
+        ):
+            paddle.pow_(var, [2])
+
```
Comment (jeff41404) [inline test/legacy_test/test_pow.py:250]: shall we check gradients of inplace `paddle.pow_`?

**Comment by jeff41404** [comment]
in description, before code of `pow_ad_func`, `auto api_result = paddle::experimental::sparse::pow(x, factor);` shall be `auto api_result = paddle::experimental::pow(x, y);` ?

File: test/legacy_test/test_pow.py (hunk 4 above)
Comment (GGBond8488) [reply test/legacy_test/test_pow.py:250]: TestInplacePowerScalar(Tensor) inherit from TestDygraphInplace,  which contains the backward test

**Comment by GGBond8488** [comment]
> in description, before code of `pow_ad_func`, `auto api_result = paddle::experimental::sparse::pow(x, factor);` shall be `auto api_result = paddle::experimental::pow(x, y);` ?

It has been changed to auto api_result = paddle::experimental::pow(x, y);

**Review by chenwhql — APPROVED** [review]

**Review by jeff41404 — APPROVED** [review]
LGTM

**Review by sunzhongkai588 — APPROVED** [review]
no doc's changes. LGTM

File: python/paddle/tensor/math.py [hunk 5]
```diff
@@ -474,6 +474,22 @@ def pow(x, y, name=None):
             )
 
 
+@inplace_apis_in_dygraph_only
+def pow_(x, y, name=None):
+    """
+    Inplace version of ``pow`` API, the output Tensor will be inplaced with input ``x``.
+    Please refer to :ref:`api_tensor_pow`.
```
Comment (sunzhongkai588) [inline python/paddle/tensor/math.py:481]: 好像这个标签不正确，导致没有正确引用。试试改成`api_paddle_pow`，或者在[api_label ](https://github.com/PaddlePaddle/docs/blob/develop/docs/api/api_label)中加入该标签
image

**Pull request merged**